x/gamm/keeper/migrate.go

```diff
@@ -0,0 +1,204 @@
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, err
+	}
+
+	// Determine the max and min ticks for the concentrated pool we are migrating to.
+	minTick, maxTick := cl.GetMinAndMaxTicksFromExponentAtPriceOne(concentratedPool.GetPrecisionFactorAtPriceOne())
+
+	// Create a full range (min to max tick) concentrated liquidity position.
+	amount0, amount1, liquidity, err = k.clKeeper.CreatePosition(ctx, poolIdEntering, sender, exitCoins.AmountOf(concentratedPool.GetToken0()), exitCoins.AmountOf(concentratedPool.GetToken1()), sdk.ZeroInt(), sdk.ZeroInt(), minTick, maxTick)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, err
+	}
+	return amount0, amount1, liquidity, poolIdLeaving, nil
+}
+
+// GetMigrationInfo returns the balancer to gamm pool migration info from the store
+// Returns an empty MigrationRecords struct if migration info does not exist
+func (k Keeper) GetMigrationInfo(ctx sdk.Context) types.MigrationRecords {
+	store := ctx.KVStore(k.storeKey)
+	migrationInfo := types.MigrationRecords{}
+	osmoutils.MustGet(store, types.KeyMigrationInfo, &migrationInfo)
+	return migrationInfo
+}
+
+// SetMigrationInfo sets the balancer to gamm pool migration info to the store
+func (k Keeper) SetMigrationInfo(ctx sdk.Context, migrationInfo types.MigrationRecords) {
+	store := ctx.KVStore(k.storeKey)
+	osmoutils.MustSet(store, types.KeyMigrationInfo, &migrationInfo)
+}
+
+// validateRecords validates a list of BalancerToConcentratedPoolLink records to ensure that:
+// 1) there are no duplicates
+// 2) both the balancer and gamm pool IDs are valid
+// It also reorders records from lowest to highest balancer pool ID if they are not provided in order already.
+func (k Keeper) validateRecords(ctx sdk.Context, records []types.BalancerToConcentratedPoolLink) error {
+	lastBalancerPoolID := uint64(0)
+	balancerIdFlags := make(map[uint64]bool, len(records))
+	clIdFlags := make(map[uint64]bool, len(records))
+
+	// Sort the provided records by balancer pool ID
+	sort.SliceStable(records, func(i, j int) bool {
+		return records[i].BalancerPoolId < records[j].BalancerPoolId
+	})
+
+	for _, record := range records {
+		// If the balancer ID has already been seen, we have a duplicate
+		if balancerIdFlags[record.BalancerPoolId] {
+			return fmt.Errorf(
+				"Balancer pool ID #%d has duplications.",
+				record.BalancerPoolId,
+			)
+		}
+
+		// If the concentrated ID has already been seen, we have a duplicate
+		if clIdFlags[record.ClPoolId] {
+			return fmt.Errorf(
+				"Concentrated pool ID #%d has duplications.",
+				record.ClPoolId,
+			)
+		}
+
+		// Ensure records are sorted from lowest to highest balancer pool ID
+		if record.BalancerPoolId < lastBalancerPoolID {
+			return fmt.Errorf(
+				"Balancer pool ID #%d came after Balancer pool ID #%d.",
+				record.BalancerPoolId, lastBalancerPoolID,
+			)
+		}
+
+		// Ensure the provided balancerPoolId exists and that it is of type balancer
+		poolType, err := k.GetPoolType(ctx, record.BalancerPoolId)
```

Comment:
Suggest changing `GetType` to be a method on the pool struct.

Then,
- get pool by id
- `GetType` on the pool

Having this defined on keeper doesn't feel right, especially considering that CL has only one type

Reply:
Agreed, addressed here! 

https://github.com/osmosis-labs/osmosis/pull/4051/commits/a35fecb25da1660673532231fdc0d5ec821c2ad2